internal/query/roles_v2.go

```diff
@@ -0,0 +1,668 @@
+		}
+
+		return role, nil
+	}
+
+	// 1. update role in permission-api DB
+	dbRole, err := e.store.UpdateRole(dbCtx, actor.ID, role.ID, newName)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return types.Role{}, err
+	}
+
+	// 2. update permissions relationships in spice db
+	updates := []*pb.RelationshipUpdate{}
+	roleRef := resourceToSpiceDBRef(e.namespace, roleResource)
+
+	// 2.a remove old actions
+	for _, action := range rmActions {
+		updates = append(
+			updates,
+			e.createRoleV2RelationshipUpdatesForAction(
+				action, roleRef,
+				pb.RelationshipUpdate_OPERATION_DELETE,
+			)...,
+		)
+	}
+
+	// 2.b add new actions
+	for _, action := range addActions {
+		updates = append(
+			updates,
+			e.createRoleV2RelationshipUpdatesForAction(
+				action, roleRef,
+				pb.RelationshipUpdate_OPERATION_TOUCH,
+			)...,
+		)
+	}
+
+	// 2.c write updates to spicedb
+	request := &pb.WriteRelationshipsRequest{Updates: updates}
+
+	if _, err := e.client.WriteRelationships(ctx, request); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return types.Role{}, err
+	}
+
+	if err = e.store.CommitContext(dbCtx); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		// At this point, spicedb changes have already been applied.
+		// Attempting to rollback could result in failures that could result in the same situation.
+		//
+		// TODO: add spicedb rollback logic along with rollback failure scenarios.
+
+		return types.Role{}, err
+	}
+
+	role.Name = dbRole.Name
+	role.CreatedBy = dbRole.CreatedBy
+	role.UpdatedBy = dbRole.UpdatedBy
+	role.ResourceID = dbRole.ResourceID
+	role.CreatedAt = dbRole.CreatedAt
+	role.UpdatedAt = dbRole.UpdatedAt
+	role.Actions = newActions
+
+	return role, nil
+}
+
+func (e *engine) DeleteRoleV2(ctx context.Context, roleResource types.Resource) error {
+	ctx, span := e.tracer.Start(ctx, "engine.DeleteRoleV2")
+	defer span.End()
+
+	dbRole, err := e.store.GetRoleByID(ctx, roleResource.ID)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	roleOwner, err := e.NewResourceFromID(dbRole.ResourceID)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	dbCtx, err := e.store.BeginContext(ctx)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	err = e.store.LockRoleForUpdate(dbCtx, roleResource.ID)
+	if err != nil {
+		sErr := fmt.Errorf("failed to lock role: %s: %w", roleResource.ID, err)
+
+		span.RecordError(sErr)
+		span.SetStatus(codes.Error, sErr.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return err
+	}
```

Comment:
It might make more sense to do this earlier, since we could in theory have two handlers attempt to get the same role and then delete it twice.